dev-tools/phpstan/tests/Extension/data/preg-match.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace PhpCsFixer\PHPStan\Tests\Extension\data\preg_match;
+
+use PhpCsFixer\Preg;
+use function PHPStan\Testing\assertType;
+
+function (string $name): void {
+    Preg::match('/^([a-z])([a-z0-9_]*)$/', $name, $matches);
+    assertType('array{0?: string, 1?: non-empty-string, 2?: string}', $matches);
+};
+
+function (string $name): void {
+    if (Preg::match('/^([a-z])([a-z0-9_]*)$/', $name, $matches)) {
+        assertType('array{string, non-empty-string, string}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+};
+
+function (string $name): void {
+    Preg::match('/^([a-z])([a-z0-9_]*)$/', $name, $matches, PREG_OFFSET_CAPTURE);
+    assertType('array{0?: array{string, int<-1, max>}, 1?: array{non-empty-string, int<-1, max>}, 2?: array{string, int<-1, max>}}', $matches);
+};
+
+function (string $name): void {
```

Comment:
please add test for https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/4a8f7b4dfa/src/DocBlock/TypeExpression.php#L434 regex

(slightly simplied) repro https://phpstan.org/r/efd74245-ebe3-4f71-b1b3-0155f05c3c71 shows the type is either not parsed completely or simplified too early, the regex is actually not that complex, it should be supported as we use it heavily across the PHP CS Fixer codebase.

Reply:
it does not make sense todo this stuff here. general type inference problems based on regex need to be done in phpstan directly.

we don't need to replicate in the php-cs-fixer repo all the cases PHPStan is able to infer.

Reply:
I would say we should have '{'.TypeExpression::REGEX_TYPES.'}' type tested here, as the regex is the most complex one in this project and when editing it, we should be instantly aware if PHPStan can analyse it fully as before, otherwise it might be very tough to understand the related PHPStan failures.

Reply:
@staabm I have discovered at least 3 currently unparsed usecases and opened an issue for them - https://github.com/phpstan/phpstan/issues/11746